scanner/engine/scheduler.h

```diff
@@ -0,0 +1,53 @@
+#pragma once
+
+#include "scanner/engine/metadata.h"
+#include "scanner/engine/table_meta_cache.h"
+#include "scanner/engine/runtime.h"
+#include "scanner/engine/dag_analysis.h"
+
+#include <deque>
+
+
+namespace scanner {
+namespace internal {
+
+class Scheduler {
+public:
+  // This function should do dependency analysis and fill up global_data_map,
+  // global_dependency_map and global_status_map.
+  void fill_global_maps(DatabaseMetadata& meta, TableMetaCache& table_metas,
+                           const std::vector<proto::Job>& jobs,
+                           const std::vector<proto::Op>& ops,
+                           DAGAnalysisInfo& info);
+
+  // If the finished op is sink, write to storage (and update global_data_map);
+  // if the finished op is not sink, update global_status_map so its successors
+  // can be assigned.
+  Result receive_finished_work();
+
+  // Give work to an actor based on the global maps we have.
+  Result give_actor_work();
+
+private:
+  // A unit is a frame on a sequence of a stage.
+  enum unit_status {
```

Comment:
We usually refer to each point in a sequence as an `Element`. Perhaps rename this to `ElementStatus`?